src/codegen/codegen_neuron_cpp_visitor.cpp

```diff
@@ -443,49 +538,78 @@ void CodegenNeuronCppVisitor::print_mechanism_register() {
 }
 
 
-void CodegenNeuronCppVisitor::print_mechanism_range_var_structure(
-    [[maybe_unused]] bool print_initializers) {
+void CodegenNeuronCppVisitor::print_mechanism_range_var_structure(bool print_initializers) {
+    auto const value_initialize = print_initializers ? "{}" : "";
+    auto int_type = default_int_data_type();
     printer->add_newline(2);
-    printer->add_line("/* NEURON RANGE variables macro definitions */");
-    for (auto i = 0; i < codegen_float_variables.size(); ++i) {
-        const auto float_var = codegen_float_variables[i];
-        if (float_var->is_array()) {
-            printer->add_line("#define ",
-                              float_var->get_name(),
-                              "(id) _ml->template data_array<",
-                              std::to_string(i),
-                              ", ",
-                              std::to_string(float_var->get_length()),
-                              ">(id)");
+    printer->add_line("/** all mechanism instance variables and global variables */");
+    printer->fmt_push_block("struct {} ", instance_struct());
+
+    for (auto const& [var, type]: info.neuron_global_variables) {
+        auto const name = var->get_name();
+        printer->fmt_line("{}* {}{};",
+                          type,
+                          name,
+                          print_initializers ? fmt::format("{{&coreneuron::{}}}", name)
+                                             : std::string{});
+    }
+    for (auto& var: codegen_float_variables) {
+        const auto& name = var->get_name();
+        // auto type = get_range_var_float_type(var);
+        // auto qualifier = is_constant_variable(name) ? "const " : "";
+        printer->fmt_line("double* {}{};", name, value_initialize);
+    }
+    for (auto& var: codegen_int_variables) {
+        const auto& name = var.symbol->get_name();
+        if (var.is_index || var.is_integer) {
+            auto qualifier = var.is_constant ? "const " : "";
+            printer->fmt_line("{}{}* {}{};", qualifier, int_type, name, value_initialize);
         } else {
-            printer->add_line("#define ",
-                              float_var->get_name(),
-                              "(id) _ml->template fpfield<",
-                              std::to_string(i),
-                              ">(id)");
+            auto qualifier = var.is_constant ? "const " : "";
+            auto type = var.is_vdata ? "void*" : default_float_data_type();
+            printer->fmt_line("{}{}* {}{};", qualifier, type, name, value_initialize);
         }
     }
+
+    // printer->fmt_line("{}* {}{};",
+    //                   global_struct(),
+    //                   naming::INST_GLOBAL_MEMBER,
+    //                   print_initializers ? fmt::format("{{&{}}}", global_struct_instance())
+    //                                      : std::string{});
```

Comment:
a nitpick for clarity: you are keeping commented-out code as those are something `todos` / `next steps` for the implementation?

Reply:
Kinda, yes. That function is almost identical to the cnrn one. Right now I don't need globals. That will change. Next mini-task is either ODEs with array range variables or with global variables